# Co-relation calculation

#### For two comparisons, we decided to calculate the co-relation of the variables being visualised using a scatter plot so as to better understand the visualisations and this would allow us to make a (_somewhat_) generalisable insight from the data. 

The function that takes a pandas DataFrame df with two columns containing data for two variables, and returns the correlation coefficient between the two variables using the pearsonr function from the scipy.stats module:

In [73]:
import pandas as pd
import scipy.stats as stats

viz1_2016 = [
    {
      "title": "Estonia",
      "id": "EE",
      "PRIS_OFF_CAP": 221.51, 
      "PRIS_ACT_CAP":217.64,
      "Living Conditions": (217.64 / 221.51),
      "GF03": 1.9
    },
    {
      "title": "Sweden",
      "id": "SE",
      "PRIS_OFF_CAP": 63.47, 
      "PRIS_ACT_CAP": 57.44,
      "Living Conditions": (57.44 / 63.47),
      "GF03": 1.3
    },
    {
      "title": "Denmark",
      "id": "DK",
      "PRIS_OFF_CAP": 61.71,
      "PRIS_ACT_CAP": 59.71,
      "Living Conditions": (61.71 / 59.71),
      "GF03": 1.0
    },
    {
      "title": "Finland",
      "id": "FI",
      "PRIS_OFF_CAP": 53.92,
      "PRIS_ACT_CAP": 57.51,
      "Living Conditions": (57.51 / 53.92),
      "GF03": 1.2
    },
    {
      "title": "Bulgaria",
      "id": "BG",
      "PRIS_OFF_CAP": 118.69, 
      "PRIS_ACT_CAP":102.67,
      "Living Conditions": (61.71 / 59.71),
      "GF03": 2.3
    },
    {
      "title": "Hungary",
      "id": "HG",
      "PRIS_OFF_CAP": 140.11, 
      "PRIS_ACT_CAP":179.62,
      "Living Conditions":(179.62 / 140.11),
      "GF03": 2.3
    },
    {
      "title": "Austria",
      "id": "AT",
      "PRIS_OFF_CAP": 101.43, 
      "PRIS_ACT_CAP":99.18,
      "Living Conditions": (99.18 / 101.43),
      "GF03": 1.3
    },
    {
      "title": "Poland",
      "id": "PL",
      "PRIS_OFF_CAP": 234.88, 
      "PRIS_ACT_CAP":190.19,
      "Living Conditions": (190.19 / 234.88),
      "GF03": 2.2
    },
    {
      "title": "Romania",
      "id": "RO",
      "PRIS_OFF_CAP": 141.13, 
      "PRIS_ACT_CAP":138.94,
      "Living Conditions": (138.94 / 141.13),
      "GF03": 2.1
    },
    {
      "title": "Slovakia",
      "id": "SK",
      "PRIS_OFF_CAP": 203.33, 
      "PRIS_ACT_CAP":184.20,
      "Living Conditions": (184.20 / 203.33),
      "GF03": 2.4
    },
    {
      "title": "Slovenia",
      "id": "SI",
      "PRIS_OFF_CAP": 64.04, 
      "PRIS_ACT_CAP":63.37,
      "Living Conditions": (63.37 / 64.04),
      "GF03": 1.7
    },
    {
      "title": "Greece",
      "id": "GR",
      "PRIS_OFF_CAP": 91.02, 
      "PRIS_ACT_CAP":88.65,
      "Living Conditions": (91.02 / 88.65), 
      "GF03": 2.2
    },
    {
      "title": "Italy",
      "id": "IT",
      "PRIS_OFF_CAP":  83.75, 
      "PRIS_ACT_CAP": 92.27,
      "Living Conditions": (92.27 /  83.75 ),
      "GF03": 1.8
    },
    {
      "title": "Portugal",
      "id": "PT",
      "PRIS_OFF_CAP": 121.84, 
      "PRIS_ACT_CAP":134.58,
      "Living Conditions": (134.58 / 121.84),
      "GF03": 1.8
    },
    {
      "title": "Spain",
      "id": "ES",
      "PRIS_OFF_CAP":   138.12, 
      "PRIS_ACT_CAP":128.31,
      "Living Conditions": (128.31 / 138.12),
      "GF03": 1.9
    },
    {
      "title": "France",
      "id": "FR",
      "PRIS_OFF_CAP": 88.06, 
      "PRIS_ACT_CAP":102.50,
      "Living Conditions": (102.50 / 88.06),
      "GF03": 1.6
    },
    {
      "title": "Germany",
      "id": "DE",
      "PRIS_OFF_CAP": 89.60, 
      "PRIS_ACT_CAP": 78.24,
      "Living Conditions": (78.24 / 89.60),
      "GF03": 1.5
    }]

viz1_2017 = [{
      "title": "Estonia",
      "id": "EE",
      "PRIS_OFF_CAP": 221.57, 
      "PRIS_ACT_CAP":206.97,
      "Living Conditions": (206.97 / 221.57) ,
      "GF03": 1.9
    },
    {
      "title": "Sweden",
      "id": "SE",
      "PRIS_OFF_CAP": 61.50, 
      "PRIS_ACT_CAP": 57.16,
      "Living Conditions": 57.16 / 61.50 ,
      "GF03": 1.3
    },
    {
      "title": "Denmark",
      "id": "DK",
      "PRIS_OFF_CAP": 62.48,
      "PRIS_ACT_CAP": 59.46,
      "Living Conditions": (59.46 /  62.48) ,
      "GF03": 0.9
    },
    {
      "title": "Finland",
      "id": "FI",
      "PRIS_OFF_CAP": 53.10,
      "PRIS_ACT_CAP": 56.00,
      "Living Conditions": (56.00 / 53.10) ,
      "GF03": 1.1
    },
    {
      "title": "Bulgaria",
      "id": "BG",
      "PRIS_OFF_CAP": 119.56, 
      "PRIS_ACT_CAP":98.40,
      "Living Conditions": (98.40 / 119.56) ,
      "GF03": 2.5
    },
    {
      "title": "Hungary",
      "id": "HG",
      "PRIS_OFF_CAP": 143.00, 
      "PRIS_ACT_CAP":177.01,
      "Living Conditions": (177.01 / 143.00) ,
      "GF03": 2.3
    },
    {
      "title": "Austria",
      "id": "AT",
      "PRIS_OFF_CAP": 99.69, 
      "PRIS_ACT_CAP":100.90,
      "Living Conditions": (100.90 / 99.69) ,
      "GF03": 1.4
    },
    {
      "title": "Poland",
      "id": "PL",
      "PRIS_OFF_CAP": 228.76, 
      "PRIS_ACT_CAP":196.14,
      "Living Conditions": (196.14 / 228.76),
      "GF03": 2.1
    },
    {
      "title": "Romania",
      "id": "RO",
      "PRIS_OFF_CAP": 103.89, 
      "PRIS_ACT_CAP":119.37,
      "Living Conditions": (119.37 / 103.89) ,
      "GF03": 2.0
    },
    {
      "title": "Slovakia",
      "id": "SK",
      "PRIS_OFF_CAP": 201.29, 
      "PRIS_ACT_CAP":184.50,
      "Living Conditions": (184.50 / 201.29) ,
      "GF03": 2.1
    },
    {
      "title": "Greece",
      "id": "GR",
      "PRIS_OFF_CAP": 92.26, 
      "PRIS_ACT_CAP":92.97,
      "Living Conditions": (92.97 / 92.26) , 
      "GF03": 2.1
    },
    {
      "title": "Italy",
      "id": "IT",
      "PRIS_OFF_CAP":  84.30, 
      "PRIS_ACT_CAP":97.44,
      "Living Conditions": (97.44 / 84.30) ,
      "GF03": 1.8
    },
    {
      "title": "Portugal",
      "id": "PT",
      "PRIS_OFF_CAP": 123.13, 
      "PRIS_ACT_CAP":131.79,
      "Living Conditions": (131.79 /  123.13) ,
      "GF03": 1.7
    },
    {
      "title": "Spain",
      "id": "ES",
      "PRIS_OFF_CAP":   137.79 , 
      "PRIS_ACT_CAP":126.40,
      "Living Conditions": (126.40 /  137.79) ,
      "GF03": 1.8
    },
    {
      "title": "France",
      "id": "FR",
      "PRIS_OFF_CAP":89.46, 
      "PRIS_ACT_CAP":103.25,
      "Living Conditions": (103.25 / 89.46) ,
      "GF03": 1.6
    },
    {
      "title": "Germany",
      "id": "DE",
      "PRIS_OFF_CAP": 89.19, 
      "PRIS_ACT_CAP": 79.79,
      "Living Conditions": (79.79 /  89.19) ,
      "GF03": 1.6
    },
    {
      "title": "Netherlands",
      "id": "NL",
      "PRIS_OFF_CAP": 70.69, 
      "PRIS_ACT_CAP":63.66,
      "Living Conditions": (63.66 / 70.69) ,
      "GF03": 1.8
    }]

viz1_2018 = [{
      "title": "Estonia",
      "id": "EE",
      "PRIS_OFF_CAP": 252.74, 
      "PRIS_ACT_CAP":195.89,
      "Living Conditions": (195.89 / 252.74) ,
      "GF03": 1.9
    },
    {
      "title": "Sweden",
      "id": "SE",
      "PRIS_OFF_CAP": 60.41, 
      "PRIS_ACT_CAP": 64.61,
      "Living Conditions": 64.61 / 60.41,
      "GF03": 1.3
    },
    {
      "title": "Denmark",
      "id": "DK",
      "PRIS_OFF_CAP": 65.64,
      "PRIS_ACT_CAP": 62.88,
      "Living Conditions": (62.88/  65.64) ,
      "GF03": 0.9
    },
    {
      "title": "Finland",
      "id": "FI",
      "PRIS_OFF_CAP": 53.96,
      "PRIS_ACT_CAP": 53.62,
      "Living Conditions": (53.62 / 53.96) ,
      "GF03": 1.1
    },
    {
      "title": "Bulgaria",
      "id": "BG",
      "PRIS_OFF_CAP": 116.89, 
      "PRIS_ACT_CAP":94.34,
      "Living Conditions": (94.34 / 116.89) ,
      "GF03": 2.5
    },
    {
      "title": "Hungary",
      "id": "HG",
      "PRIS_OFF_CAP": 144.70, 
      "PRIS_ACT_CAP":177.01,
      "Living Conditions": (177.01 / 144.70) ,
      "GF03": 2.2
    },
    {
      "title": "Austria",
      "id": "AT",
      "PRIS_OFF_CAP": 100.24, 
      "PRIS_ACT_CAP":103.86,
      "Living Conditions": (103.86 / 100.24) ,
      "GF03": 1.4
    },
    {
      "title": "Poland",
      "id": "PL",
      "PRIS_OFF_CAP": 226.30, 
      "PRIS_ACT_CAP":191.74 ,
      "Living Conditions": (196.14 / 228.76),
      "GF03": 2.1
    },
    {
      "title": "Romania",
      "id": "RO",
      "PRIS_OFF_CAP": 99.39, 
      "PRIS_ACT_CAP":106.46,
      "Living Conditions": (106.46 / 99.39) ,
      "GF03": 2.2
    },
    {
      "title": "Slovakia",
      "id": "SK",
      "PRIS_OFF_CAP": 211.26, 
      "PRIS_ACT_CAP":189.12,
      "Living Conditions": (189.12 / 211.26) ,
      "GF03": 2.2
    },
    {
      "title": "Greece",
      "id": "GR",
      "PRIS_OFF_CAP": 92.49, 
      "PRIS_ACT_CAP":99.19,
      "Living Conditions": (99.19 / 92.49) , 
      "GF03": 2.3
    },
    {
      "title": "Italy",
      "id": "IT",
      "PRIS_OFF_CAP":  84.55 , 
      "PRIS_ACT_CAP":101.07,
      "Living Conditions": (101.07/ 84.55 ) ,
      "GF03": 1.8
    },
    {
      "title": "Portugal",
      "id": "PT",
      "PRIS_OFF_CAP": 125.68, 
      "PRIS_ACT_CAP":126.53 ,
      "Living Conditions": (131.79 /  125.68) ,
      "GF03": 1.7
    },
    {
      "title": "Spain",
      "id": "ES",
      "PRIS_OFF_CAP":137.67, 
      "PRIS_ACT_CAP":126.20,
      "Living Conditions": (126.20 /  137.67) ,
      "GF03": 1.8
    },
    {
      "title": "France",
      "id": "FR",
      "PRIS_OFF_CAP":89.74, 
      "PRIS_ACT_CAP":104.69,
      "Living Conditions": (104.69 / 89.74) ,
      "GF03": 1.6
    },
    {
      "title": "Germany",
      "id": "DE",
      "PRIS_OFF_CAP": 89.85, 
      "PRIS_ACT_CAP": 79.43 ,
      "Living Conditions": (79.43 / 89.85) ,
      "GF03": 1.6
    },
    {
      "title": "Netherlands",
      "id": "NL",
      "PRIS_OFF_CAP": 72.03, 
      "PRIS_ACT_CAP":65.48,
      "Living Conditions": (65.48 / 72.03) ,
      "GF03": 1.8
    }]

viz1_2019 = [{
      "title": "Estonia",
      "id": "EE",
      "PRIS_OFF_CAP": 247.43, 
      "PRIS_ACT_CAP":187.72,
      "Living Conditions": (187.72 / 247.43) ,
      "GF03": 1.8
    },
    {
      "title": "Sweden",
      "id": "SE",
      "PRIS_OFF_CAP": 65.07 , 
      "PRIS_ACT_CAP": 65.68 ,
      "Living Conditions": (65.68 / 65.07),
      "GF03": 1.3
    },
    {
      "title": "Denmark",
      "id": "DK",
      "PRIS_OFF_CAP": 67.98,
      "PRIS_ACT_CAP": 67.52,
      "Living Conditions": (67.52/67.98) ,
      "GF03": 1.0
    },
    {
      "title": "Finland",
      "id": "FI",
      "PRIS_OFF_CAP": 52.16,
      "PRIS_ACT_CAP": 53.37 ,
      "Living Conditions": (53.37  / 52.16) ,
      "GF03": 1.2
    },
    {
      "title": "Bulgaria",
      "id": "BG",
      "PRIS_OFF_CAP": 115.79 , 
      "PRIS_ACT_CAP":92.11,
      "Living Conditions": (92.11/ 115.79 ) ,
      "GF03": 2.7
    },
    {
      "title": "Hungary",
      "id": "HG",
      "PRIS_OFF_CAP": 152.12, 
      "PRIS_ACT_CAP":170.51,
      "Living Conditions": (170.51 / 152.12) ,
      "GF03": 2.1
    },
    {
      "title": "Austria",
      "id": "AT",
      "PRIS_OFF_CAP": 100.27, 
      "PRIS_ACT_CAP":102.41,
      "Living Conditions": (102.41 / 100.27) ,
      "GF03": 1.3
    },
    {
      "title": "Poland",
      "id": "PL",
      "PRIS_OFF_CAP": 225.82, 
      "PRIS_ACT_CAP":196.49  ,
      "Living Conditions": (196.49/225.82),
      "GF03": 2.1
    },
    {
      "title": "Romania",
      "id": "RO",
      "PRIS_OFF_CAP": 100.55, 
      "PRIS_ACT_CAP":105.99,
      "Living Conditions": (105.99 / 100.55) ,
      "GF03": 2.2
    },
    {
      "title": "Slovakia",
      "id": "SK",
      "PRIS_OFF_CAP": 213.29, 
      "PRIS_ACT_CAP":193.65,
      "Living Conditions": (193.65 / 213.29) ,
      "GF03": 2.3
    },
    {
      "title": "Greece",
      "id": "GR",
      "PRIS_OFF_CAP": 92.64 , 
      "PRIS_ACT_CAP":101.55 ,
      "Living Conditions": (101.55  / 92.64 ) , 
      "GF03": 2.1
    },
    {
      "title": "Italy",
      "id": "IT",
      "PRIS_OFF_CAP":  85.68 , 
      "PRIS_ACT_CAP":104.04 ,
      "Living Conditions": (104.04 / 85.68 ) ,
      "GF03": 1.8
    },
    {
      "title": "Portugal",
      "id": "PT",
      "PRIS_OFF_CAP": 125.86 , 
      "PRIS_ACT_CAP":125.99 ,
      "Living Conditions": (125.99/  125.86 ) ,
      "GF03": 1.6
    },
    {
      "title": "Spain",
      "id": "ES",
      "PRIS_OFF_CAP":136.85, 
      "PRIS_ACT_CAP":124.67,
      "Living Conditions": (124.67 /  136.85) ,
      "GF03": 1.8
    },
    {
      "title": "France",
      "id": "FR",
      "PRIS_OFF_CAP":90.86 , 
      "PRIS_ACT_CAP":105.30,
      "Living Conditions": (105.30 / 90.86 ) ,
      "GF03": 1.6
    },
    {
      "title": "Germany",
      "id": "DE",
      "PRIS_OFF_CAP": 87.89 , 
      "PRIS_ACT_CAP": 76.06 ,
      "Living Conditions": (76.06 / 87.89 ) ,
      "GF03": 1.6
    },
    {
      "title": "Netherlands",
      "id": "NL",
      "PRIS_OFF_CAP": 67.20 , 
      "PRIS_ACT_CAP":68.15 ,
      "Living Conditions": (68.15 / 67.20) ,
      "GF03": 1.8
    }]

viz1_2020 = [{
      "title": "Estonia",
      "id": "EE",
      "PRIS_OFF_CAP": 246.66, 
      "PRIS_ACT_CAP":178.03,
      "Living Conditions": (178.03 / 246.66) ,
      "GF03": 2.0
    },
    {
      "title": "Sweden",
      "id": "SE",
      "PRIS_OFF_CAP": 67.38 , 
      "PRIS_ACT_CAP": 70.66  ,
      "Living Conditions": (70.66  / 67.38),
      "GF03": 1.4
    },
    {
      "title": "Denmark",
      "id": "DK",
      "PRIS_OFF_CAP": 70.33,
      "PRIS_ACT_CAP": 71.55,
      "Living Conditions": (71.55/70.33) ,
      "GF03": 1.0
    },
    {
      "title": "Finland",
      "id": "FI",
      "PRIS_OFF_CAP": 54.24,
      "PRIS_ACT_CAP": 49.79,
      "Living Conditions": (49.79/54.24) ,
      "GF03": 1.2
    },
    {
      "title": "Bulgaria",
      "id": "BG",
      "PRIS_OFF_CAP": 117.40, 
      "PRIS_ACT_CAP":89.92 ,
      "Living Conditions": (89.92 / 117.40 ) ,
      "GF03": 2.7
    },
    {
      "title": "Hungary",
      "id": "HG",
      "PRIS_OFF_CAP":162.80 , 
      "PRIS_ACT_CAP":171.21,
      "Living Conditions": (171.21/ 162.80) ,
      "GF03": 2.1
    },
    {
      "title": "Austria",
      "id": "AT",
      "PRIS_OFF_CAP": 99.64 , 
      "PRIS_ACT_CAP":95.36,
      "Living Conditions": (95.36 / 99.64 ) ,
      "GF03": 1.4
    },
    {
      "title": "Poland",
      "id": "PL",
      "PRIS_OFF_CAP": 226.75 , 
      "PRIS_ACT_CAP":180.08,
      "Living Conditions": (180.08/226.75 ),
      "GF03": 2.3
    },
    {
      "title": "Romania",
      "id": "RO",
      "PRIS_OFF_CAP": 101.05, 
      "PRIS_ACT_CAP":112.54,
      "Living Conditions": (105.99 / 101.05) ,
      "GF03": 2.5
    },
    {
      "title": "Slovakia",
      "id": "SK",
      "PRIS_OFF_CAP": 213.00 , 
      "PRIS_ACT_CAP":192.73 ,
      "Living Conditions": (192.73 / 213.00) ,
      "GF03": 2.4
    },
    {
      "title": "Greece",
      "id": "GR",
      "PRIS_OFF_CAP": 94.93, 
      "PRIS_ACT_CAP":106.16,
      "Living Conditions": (106.16/ 94.93) , 
      "GF03": 2.4
    },
    {
      "title": "Italy",
      "id": "IT",
      "PRIS_OFF_CAP":  85.72, 
      "PRIS_ACT_CAP":91.56 ,
      "Living Conditions": (91.56  / 85.72) ,
      "GF03": 2.0
    },
    {
      "title": "Portugal",
      "id": "PT",
      "PRIS_OFF_CAP": 125.52 , 
      "PRIS_ACT_CAP":111.71,
      "Living Conditions": (111.71/125.52 ) ,
      "GF03": 1.9
    },
    {
      "title": "Spain",
      "id": "ES",
      "PRIS_OFF_CAP":136.13, 
      "PRIS_ACT_CAP":116.58,
      "Living Conditions": (116.58/ 136.13) ,
      "GF03": 2.1
    },
    {
      "title": "France",
      "id": "FR",
      "PRIS_OFF_CAP":89.99, 
      "PRIS_ACT_CAP":93.10,
      "Living Conditions": (93.10 / 89.99 ) ,
      "GF03": 1.8
    },
    {
      "title": "Germany",
      "id": "DE",
      "PRIS_OFF_CAP": 87.00 , 
      "PRIS_ACT_CAP": 71.65,
      "Living Conditions": (71.65/87.00 ) ,
      "GF03": 1.7
    },
    {
      "title": "Netherlands",
      "id": "NL",
      "PRIS_OFF_CAP": 69.49 , 
      "PRIS_ACT_CAP":63.33,
      "Living Conditions": (63.33 / 69.49) ,
      "GF03": 2.0
    }]
# storing the data in a df
DF1 = pd.DataFrame() 
DF2 = pd.DataFrame()
DF3 = pd.DataFrame() 
DF4 = pd.DataFrame()
DF5 = pd.DataFrame()
def createDF(data,df):
    for item in data:
        s = pd.Series(item)
        df = pd.concat([df, s.to_frame().T], ignore_index=True)

    return df

DF1 = createDF(viz1_2016,DF1)    
DF2 = createDF(viz1_2017,DF2)
DF3 = createDF(viz1_2018,DF3)
DF4 = createDF(viz1_2019,DF4)
DF5 = createDF(viz1_2020,DF5)
DF1 = DF1.drop(columns=['PRIS_OFF_CAP','id','PRIS_ACT_CAP','title'])
DF2 = DF2.drop(columns=['PRIS_OFF_CAP','id','PRIS_ACT_CAP','title'])
DF3 = DF3.drop(columns=['PRIS_OFF_CAP','id','PRIS_ACT_CAP','title'])
DF4 = DF4.drop(columns=['PRIS_OFF_CAP','id','PRIS_ACT_CAP','title'])
DF5 = DF5.drop(columns=['PRIS_OFF_CAP','id','PRIS_ACT_CAP','title'])


# ====================================================================================================

In [76]:
# data for the second plot chart

viz5_2016 = [{
      
        "title": "Estonia",
        "id": "EE",
        "PRIS_OFF_CAP": 221.51, 
        "PRIS_ACT_CAP":217.64,
        "Living Conditions": (217.64 / 221.51),
      },
      {
        "title": "Sweden",
        "id": "SE",
        "PRIS_OFF_CAP": 63.47, 
        "PRIS_ACT_CAP": 57.44,
        "Living Conditions": (57.44 / 63.47),
      },
      {
        "title": "Denmark",
        "id": "DK",
        "PRIS_OFF_CAP": 61.71,
        "PRIS_ACT_CAP": 59.71,
        "Living Conditions": (61.71 / 59.71),
        "Suicides": 8 
      },
      {
        "title": "Finland",
        "id": "FI",
        "PRIS_OFF_CAP": 53.92,
        "PRIS_ACT_CAP": 57.51,
        "Living Conditions": (57.51 / 53.92),
      },
      {
        "title": "Bulgaria",
        "id": "BG",
        "PRIS_OFF_CAP": 118.69, 
        "PRIS_ACT_CAP":102.67,
        "Living Conditions": (61.71 / 59.71),
        "Suicides": 3
      },
      {
        "title": "Hungary",
        "id": "HG",
        "PRIS_OFF_CAP": 140.11, 
        "PRIS_ACT_CAP":179.62,
        "Living Conditions":(179.62 / 140.11),
        "Suicides": 9
      },
      {
        "title": "Austria",
        "id": "AT",
        "PRIS_OFF_CAP": 101.43, 
        "PRIS_ACT_CAP":99.18,
        "Living Conditions": (99.18 / 101.43),
        "Suicides": 10
      },
      {
        "title": "Poland",
        "id": "PL",
        "PRIS_OFF_CAP": 234.88, 
        "PRIS_ACT_CAP":190.19,
        "Living Conditions": (190.19 / 234.88),
        "Suicides": 24
      },
      {
        "title": "Romania",
        "id": "RO",
        "PRIS_OFF_CAP": 141.13, 
        "PRIS_ACT_CAP":138.94,
        "Living Conditions": (138.94 / 141.13),
        "Suicides": 11
      },
      {
        "title": "Slovakia",
        "id": "SK",
        "PRIS_OFF_CAP": 203.33, 
        "PRIS_ACT_CAP":184.20,
        "Living Conditions": (184.20 / 203.33),
        "Suicides": 6
      },
      {
        "title": "Slovenia",
        "id": "SI",
        "PRIS_OFF_CAP": 64.04, 
        "PRIS_ACT_CAP":63.37,
        "Living Conditions": (63.37 / 64.04),
        "Suicides": 3 
      },
      {
        "title": "Greece",
        "id": "GR",
        "PRIS_OFF_CAP": 91.02, 
        "PRIS_ACT_CAP":88.65,
        "Living Conditions": (91.02 / 88.65), 
        "Suicides": 6
      },
      {
        "title": "Italy",
        "id": "IT",
        "PRIS_OFF_CAP":  83.75, 
        "PRIS_ACT_CAP": 92.27,
        "Living Conditions": (92.27 /  83.75 ),
        "Suicides": 39
      },
      {
        "title": "Portugal",
        "id": "PT",
        "PRIS_OFF_CAP": 121.84, 
        "PRIS_ACT_CAP":134.58,
        "Living Conditions": (134.58 / 121.84),
        "Suicides": 9 
      },
      {
        "title": "Spain",
        "id": "ES",
        "PRIS_OFF_CAP":   138.12, 
        "PRIS_ACT_CAP":128.31,
        "Living Conditions": (128.31 / 138.12),
        "Suicides": 23
      },
      {
        "title": "France",
        "id": "FR",
        "PRIS_OFF_CAP": 88.06, 
        "PRIS_ACT_CAP":102.50,
        "Living Conditions": (102.50 / 88.06),
        "Suicides": 125
      },
      {
        "title": "Germany",
        "id": "DE",
        "PRIS_OFF_CAP": 89.60, 
        "PRIS_ACT_CAP": 78.24,
        "Living Conditions": (78.24 / 89.60),
        "Suicides": 76
      }]
viz5_2017 = [{
        
            "title": "Estonia",
            "id": "EE",
            "PRIS_OFF_CAP": 221.57, 
            "PRIS_ACT_CAP":206.97,
            "Living Conditions": (206.97 / 221.57) ,
            "Suicides": 1
          },
          {
            "title": "Sweden",
            "id": "SE",
            "PRIS_OFF_CAP": 61.50, 
            "PRIS_ACT_CAP": 57.16,
            "Living Conditions": 57.16 / 61.50 ,
            "Suicides": 1
    
          },
          {
            "title": "Denmark",
            "id": "DK",
            "PRIS_OFF_CAP": 62.48,
            "PRIS_ACT_CAP": 59.46,
            "Living Conditions": (59.46 /  62.48) ,
            "Suicides": 4
    
          },
          {
            "title": "Finland",
            "id": "FI",
            "PRIS_OFF_CAP": 53.10,
            "PRIS_ACT_CAP": 56.00,
            "Living Conditions": (56.00 / 53.10) ,
            "Suicides": 2
    
          },
          {
            "title": "Bulgaria",
            "id": "BG",
            "PRIS_OFF_CAP": 119.56, 
            "PRIS_ACT_CAP":98.40,
            "Living Conditions": (98.40 / 119.56) ,
            "Suicides": 5
    
          },
          {
            "title": "Hungary",
            "id": "HG",
            "PRIS_OFF_CAP": 143.00, 
            "PRIS_ACT_CAP":177.01,
            "Living Conditions": (177.01 / 143.00) ,
            "Suicides": 4
    
          },
          {
            "title": "Austria",
            "id": "AT",
            "PRIS_OFF_CAP": 99.69, 
            "PRIS_ACT_CAP":100.90,
            "Living Conditions": (100.90 / 99.69) ,
            "Suicides": 11
    
          },
          {
            "title": "Poland",
            "id": "PL",
            "PRIS_OFF_CAP": 228.76, 
            "PRIS_ACT_CAP":196.14,
            "Living Conditions": (196.14 / 228.76),
    
          },
          {
            "title": "Romania",
            "id": "RO",
            "PRIS_OFF_CAP": 103.89, 
            "PRIS_ACT_CAP":119.37,
            "Living Conditions": (119.37 / 103.89) ,
            "Suicides": 9
    
          },
          {
            "title": "Slovakia",
            "id": "SK",
            "PRIS_OFF_CAP": 201.29, 
            "PRIS_ACT_CAP":184.50,
            "Living Conditions": (184.50 / 201.29) ,
            "Suicides": 10
    
          },
          {
            "title": "Greece",
            "id": "GR",
            "PRIS_OFF_CAP": 92.26, 
            "PRIS_ACT_CAP":92.97,
            "Living Conditions": (92.97 / 92.26) , 
            "Suicides": 0
    
          },
          {
            "title": "Italy",
            "id": "IT",
            "PRIS_OFF_CAP":  84.30, 
            "PRIS_ACT_CAP":97.44,
            "Living Conditions": (97.44 / 84.30) ,
            "Suicides": 48
    
          },
          {
            "title": "Portugal",
            "id": "PT",
            "PRIS_OFF_CAP": 123.13, 
            "PRIS_ACT_CAP":131.79,
            "Living Conditions": (131.79 /  123.13) ,
            "Suicides": 15
    
          },
          {
            "title": "Spain",
            "id": "ES",
            "PRIS_OFF_CAP":   137.79 , 
            "PRIS_ACT_CAP":126.40,
            "Living Conditions": (126.40 /  137.79) ,
            "Suicides": 27
    
          },
          {
            "title": "France",
            "id": "FR",
            "PRIS_OFF_CAP":89.46, 
            "PRIS_ACT_CAP":103.25,
            "Living Conditions": (103.25 / 89.46) ,
            "Suicides": 88
    
          },
          {
            "title": "Germany",
            "id": "DE",
            "PRIS_OFF_CAP": 89.19, 
            "PRIS_ACT_CAP": 79.79,
            "Living Conditions": (79.79 /  89.19) ,
            "Suicides": 82
    
          },
          {
            "title": "Netherlands",
            "id": "NL",
            "PRIS_OFF_CAP": 70.69, 
            "PRIS_ACT_CAP":63.66,
            "Living Conditions": (63.66 / 70.69) ,
            "Suicides": 14
          }]
viz5_2018 = [{
      
        "title": "Estonia",
        "id": "EE",
        "PRIS_OFF_CAP": 252.74, 
        "PRIS_ACT_CAP":195.89,
        "Living Conditions": (195.89 / 252.74) ,       
        "Suicides": 4
      },
      {
        "title": "Sweden",
        "id": "SE",
        "PRIS_OFF_CAP": 60.41, 
        "PRIS_ACT_CAP": 64.61,
        "Living Conditions": 64.61 / 60.41,
        "Suicides": 2
      },    
      {
        "title": "Denmark",
        "id": "DK",
        "PRIS_OFF_CAP": 65.64,
        "PRIS_ACT_CAP": 62.88,
        "Living Conditions": (62.88/  65.64) ,
        "Suicides":5 },
      {
        "title": "Finland",
        "id": "FI",
        "PRIS_OFF_CAP": 53.96,
        "PRIS_ACT_CAP": 53.62,
        "Living Conditions": (53.62 / 53.96) ,
        "Suicides": 2 },
      {
        "title": "Bulgaria",
        "id": "BG",
        "PRIS_OFF_CAP": 116.89, 
        "PRIS_ACT_CAP":94.34,
        "Living Conditions": (94.34 / 116.89) ,
        "Suicides": 3 },
      {
        "title": "Hungary",
        "id": "HG",
        "PRIS_OFF_CAP": 144.70, 
        "PRIS_ACT_CAP":177.01,
        "Living Conditions": (177.01 / 144.70) ,  
        "Suicides": 6     
      },
      {
        "title": "Austria",
        "id": "AT",
        "PRIS_OFF_CAP": 100.24, 
        "PRIS_ACT_CAP":103.86,
        "Living Conditions": (103.86 / 100.24) ,      
        "Suicides": 12
      },
      {
        "title": "Poland",
        "id": "PL",
        "PRIS_OFF_CAP": 226.30, 
        "PRIS_ACT_CAP":191.74 ,
        "Living Conditions": (196.14 / 228.76),
        "Suicides": 26
     },
      {
        "title": "Romania",
        "id": "RO",
        "PRIS_OFF_CAP": 99.39, 
        "PRIS_ACT_CAP":106.46,
        "Living Conditions": (106.46 / 99.39) ,
        "Suicides": 11
     },
      {
        "title": "Slovakia",
        "id": "SK",
        "PRIS_OFF_CAP": 211.26, 
        "PRIS_ACT_CAP":189.12,
        "Living Conditions": (189.12 / 211.26) ,
        "Suicides": 5       
      },
      {
        "title": "Greece",
        "id": "GR",
        "PRIS_OFF_CAP": 92.49, 
        "PRIS_ACT_CAP":99.19,
        "Living Conditions": (99.19 / 92.49) , 
        "Suicides": 9  },
      {
        "title": "Italy",
        "id": "IT",
        "PRIS_OFF_CAP":  84.55 , 
        "PRIS_ACT_CAP":101.07,
        "Living Conditions": (101.07/ 84.55 ) ,
        "Suicides": 61 },
      {
        "title": "Portugal",
        "id": "PT",
        "PRIS_OFF_CAP": 125.68, 
        "PRIS_ACT_CAP":126.53 ,
        "Living Conditions": (131.79 /  125.68),
        "Suicides":11
      },
      {
        "title": "Spain",
        "id": "ES",
        "PRIS_OFF_CAP":137.67, 
        "PRIS_ACT_CAP":126.20,
        "Living Conditions": (126.20 /  137.67),
        "Suicides":33       
      },
      {
        "title": "France",
        "id": "FR",
        "PRIS_OFF_CAP":89.74, 
        "PRIS_ACT_CAP":104.69,
        "Living Conditions": (104.69 / 89.74) ,
        "Suicides": 103  },
      {
        "title": "Netherlands",
        "id": "NL",
        "PRIS_OFF_CAP": 72.03, 
        "PRIS_ACT_CAP":65.48,
        "Living Conditions": (65.48 / 72.03) ,
        "Suicides": 13
    }]
viz5_2019 = [{
            "title": "Estonia",
            "id": "EE",
            "PRIS_OFF_CAP": 247.43, 
            "PRIS_ACT_CAP":187.72,
            "Living Conditions": (187.72 / 247.43) ,
            "Suicides": 4
          },
          {
            "title": "Sweden",
            "id": "SE",
            "PRIS_OFF_CAP": 65.07 , 
            "PRIS_ACT_CAP": 65.68 ,
            "Living Conditions": (65.68 / 65.07),
            "Suicides": 2
          },
          {
            "title": "Denmark",
            "id": "DK",
            "PRIS_OFF_CAP": 67.98,
            "PRIS_ACT_CAP": 67.52,
            "Living Conditions": (67.52/67.98) ,
            "Suicides":9
          },
          {
            "title": "Finland",
            "id": "FI",
            "PRIS_OFF_CAP": 52.16,
            "PRIS_ACT_CAP": 53.37 ,
            "Living Conditions": (53.37  / 52.16) ,
            "Suicides": 1
          },
          {
            "title": "Bulgaria",
            "id": "BG",
            "PRIS_OFF_CAP": 115.79 , 
            "PRIS_ACT_CAP":92.11,
            "Living Conditions": (92.11/ 115.79 ) ,
            "Suicides": 3
          },
          {
            "title": "Hungary",
            "id": "HG",
            "PRIS_OFF_CAP": 152.12, 
            "PRIS_ACT_CAP":170.51,
            "Living Conditions": (170.51 / 152.12) ,
            "Suicides": 6
          },
          {
            "title": "Austria",
            "id": "AT",
            "PRIS_OFF_CAP": 100.27, 
            "PRIS_ACT_CAP":102.41,
            "Living Conditions": (102.41 / 100.27) ,
            "Suicides": 9
          },
          {
            "title": "Poland",
            "id": "PL",
            "PRIS_OFF_CAP": 225.82, 
            "PRIS_ACT_CAP":196.49  ,
            "Living Conditions": (196.49/225.82),
            "Suicides": 24
          },
          {
            "title": "Romania",
            "id": "RO",
            "PRIS_OFF_CAP": 100.55, 
            "PRIS_ACT_CAP":105.99,
            "Living Conditions": (105.99 / 100.55) ,
            "Suicides": 7
          },
          {
            "title": "Slovakia",
            "id": "SK",
            "PRIS_OFF_CAP": 213.29, 
            "PRIS_ACT_CAP":193.65,
            "Living Conditions": (193.65 / 213.29) ,
          },
          {
            "title": "Greece",
            "id": "GR",
            "PRIS_OFF_CAP": 92.64 , 
            "PRIS_ACT_CAP":101.55 ,
            "Living Conditions": (101.55  / 92.64 ) , 
          },
          {
            "title": "Italy",
            "id": "IT",
            "PRIS_OFF_CAP":  85.68 , 
            "PRIS_ACT_CAP":104.04 ,
            "Living Conditions": (104.04 / 85.68 ) ,
            "Suicides": 53
          },
          {
            "title": "Portugal",
            "id": "PT",
            "PRIS_OFF_CAP": 125.86 , 
            "PRIS_ACT_CAP":125.99 ,
            "Living Conditions": (125.99/  125.86 ) ,
            "Suicides": 11
          },
          {
            "title": "Spain",
            "id": "ES",
            "PRIS_OFF_CAP":136.85, 
            "PRIS_ACT_CAP":124.67,
            "Living Conditions": (124.67 /  136.85) ,
            "Suicides": 43
          },
          {
            "title": "France",
            "id": "FR",
            "PRIS_OFF_CAP":90.86 , 
            "PRIS_ACT_CAP":105.30,
            "Living Conditions": (105.30 / 90.86 ) ,
            "Suicides": 128
          },
          {
            "title": "Germany",
            "id": "DE",
            "PRIS_OFF_CAP": 87.89 , 
            "PRIS_ACT_CAP": 76.06 ,
            "Living Conditions": (76.06 / 87.89 ) ,
          },
          {
            "title": "Netherlands",
            "id": "NL",
            "PRIS_OFF_CAP": 67.20 , 
            "PRIS_ACT_CAP":68.15 ,
            "Living Conditions": (68.15 / 67.20) ,
            "Suicides": 12
      
        
    }]
viz5_2020 = [{
        "title": "Estonia",
        "id": "EE",
        "PRIS_OFF_CAP": 246.66, 
        "PRIS_ACT_CAP":178.03,
        "Living Conditions": (178.03 / 246.66),
        "Suicides": 3
      },
      {
        "title": "Sweden",
        "id": "SE",
        "PRIS_OFF_CAP": 67.38 , 
        "PRIS_ACT_CAP": 70.66  ,
        "Living Conditions": (70.66  / 67.38),
        "Suicides": 2
      },
      {
        "title": "Denmark",
        "id": "DK",
        "PRIS_OFF_CAP": 70.33,
        "PRIS_ACT_CAP": 71.55,
        "Living Conditions": (71.55/70.33) ,
        "Suicides": 4
      },
      {
        "title": "Finland",
        "id": "FI",
        "PRIS_OFF_CAP": 54.24,
        "PRIS_ACT_CAP": 49.79,
        "Living Conditions": (49.79/54.24) ,
        "Suicides": 4
      },
      {
        "title": "Bulgaria",
        "id": "BG",
        "PRIS_OFF_CAP": 117.40, 
        "PRIS_ACT_CAP":89.92 ,
        "Living Conditions": (89.92 / 117.40 ) ,
        "Suicides": 1
      },
      {
        "title": "Hungary",
        "id": "HG",
        "PRIS_OFF_CAP":162.80 , 
        "PRIS_ACT_CAP":171.21,
        "Living Conditions": (171.21/ 162.80) ,
        "Suicides": 5
      },
      {
        "title": "Austria",
        "id": "AT",
        "PRIS_OFF_CAP": 99.64 , 
        "PRIS_ACT_CAP":95.36,
        "Living Conditions": (95.36 / 99.64 ) ,
        "Suicides": 8
      },
      {
        "title": "Poland",
        "id": "PL",
        "PRIS_OFF_CAP": 226.75 , 
        "PRIS_ACT_CAP":180.08,
        "Living Conditions": (180.08/226.75 ),
        "Suicides": 27
      },
      {
        "title": "Romania",
        "id": "RO",
        "PRIS_OFF_CAP": 101.05, 
        "PRIS_ACT_CAP":112.54,
        "Living Conditions": (105.99 / 101.05) ,
        "Suicides": 14
      },
      {
        "title": "Slovakia",
        "id": "SK",
        "PRIS_OFF_CAP": 213.00 , 
        "PRIS_ACT_CAP":192.73 ,
        "Living Conditions": (192.73 / 213.00) ,
      },
      {
        "title": "Greece",
        "id": "GR",
        "PRIS_OFF_CAP": 94.93, 
        "PRIS_ACT_CAP":106.16,
        "Living Conditions": (106.16/ 94.93) , 
        "Suicides":6
      },
      {
        "title": "Italy",
        "id": "IT",
        "PRIS_OFF_CAP":  85.72, 
        "PRIS_ACT_CAP":91.56 ,
        "Living Conditions": (91.56  / 85.72) ,
        "Suicides":61
      },
      {
        "title": "Portugal",
        "id": "PT",
        "PRIS_OFF_CAP": 125.52 , 
        "PRIS_ACT_CAP":111.71,
        "Living Conditions": (111.71/125.52 ) ,
        "Suicides": 21
      },
      {
        "title": "Spain",
        "id": "ES",
        "PRIS_OFF_CAP":136.13, 
        "PRIS_ACT_CAP":116.58,
        "Living Conditions": (116.58/ 136.13) ,
        "Suicides":51
      },
      {
        "title": "France",
        "id": "FR",
        "PRIS_OFF_CAP":89.99, 
        "PRIS_ACT_CAP":93.10,
        "Living Conditions": (93.10 / 89.99 ) ,
        "Suicides": 119
      },
      {
        "title": "Germany",
        "id": "DE",
        "PRIS_OFF_CAP": 87.00 , 
        "PRIS_ACT_CAP": 71.65,
        "Living Conditions": (71.65/87.00 ) ,
      },
      {
        "title": "Netherlands",
        "id": "NL",
        "PRIS_OFF_CAP": 69.49 , 
        "PRIS_ACT_CAP":63.33,
        "Living Conditions": (63.33 / 69.49) ,
        "Suicides": 13
      }]

def drop_nan_rows(df):
    """
    Drop rows that contain NaN values in a Pandas DataFrame.
    """
    return df.dropna(axis=0, how='any')

DF6 = pd.DataFrame()
DF7 = pd.DataFrame()
DF8 = pd.DataFrame()
DF9 = pd.DataFrame()
DF10 = pd.DataFrame()

DF6 = createDF(viz5_2016,DF6)
DF7 = createDF(viz5_2017,DF7)
DF8 = createDF(viz5_2018,DF8)
DF9 = createDF(viz5_2019,DF9)
DF10 = createDF(viz5_2020,DF10)

DF6 = DF6.drop(columns=['title','PRIS_ACT_CAP','PRIS_OFF_CAP','id','title'])
DF7 = DF7.drop(columns=['title','PRIS_ACT_CAP','PRIS_OFF_CAP','id','title'])
DF8 = DF8.drop(columns=['title','PRIS_ACT_CAP','PRIS_OFF_CAP','id','title'])
DF9 = DF9.drop(columns=['title','PRIS_ACT_CAP','PRIS_OFF_CAP','id','title'])
DF10 = DF10.drop(columns=['title','PRIS_ACT_CAP','PRIS_OFF_CAP','id','title'])

# drop NaN
DF6 = drop_nan_rows(DF6)
DF7 = drop_nan_rows(DF7)
DF8 = drop_nan_rows(DF8)
DF9 = drop_nan_rows(DF9)
DF10 = drop_nan_rows(DF10)

print(DF9)

   Living Conditions Suicides
0           0.758679        4
1           1.009375        2
2           0.993233        9
3           1.023198        1
4           0.795492        3
5           1.120891        6
6           1.021342        9
7           0.870118       24
8           1.054102        7
11          1.214286       53
12          1.001033       11
13          0.910997       43
14          1.158926      128
16          1.014137       12


In [89]:
def calculate_correlation(df):
    """
    Calculate correlation coefficient between two variables in a pandas DataFrame.
    
    Parameters:
        df (pandas.DataFrame): DataFrame containing data for two variables.
        
    Returns:
        float: Pearson correlation coefficient between the two variables.
    """
    x = df.iloc[:,0]
    y = df.iloc[:,1]
    corr_coef, p_value = stats.pearsonr(x, y)
    return corr_coef,p_value

In [90]:
# example data
data = {'x': [1, 2, 3, 4, 5],
        'y': [2, 4, 6, 8, 10]}
df = pd.DataFrame(data)

# calculate correlation coefficient
corr_coef = calculate_correlation(DF1)
print(corr_coef)
corr_coef2 = calculate_correlation(DF2)
print(corr_coef2)
corr_coef3 = calculate_correlation(DF3)
print(corr_coef3)
corr_coef4 = calculate_correlation(DF4)
print(corr_coef4)
corr_coef5 = calculate_correlation(DF5)
print(corr_coef5)

(0.027442531754036645, 0.9167338235525603)
(-0.03630177486306102, 0.8899876254318758)
(-0.11874533301732573, 0.6498873117380816)
(-0.2440835680360957, 0.3451113199197201)
(-0.18749908998087922, 0.4711413350116708)


The first year(2016) = 0.027442531754036645 <br>
The second year(2017) = -0.03630177486306102 <br>
The third year(2018) = -0.11874533301732573 <br>
The fourth year(2019) = -0.2440835680360957 <br>
The fourth year(2020) = -0.18749908998087922 <br>


The first two year(2016,2017) = -0.003136083759218096 <br>
The first three years(2016,17,18) = -0.046247764089734876 <br>


A correlation coefficient measures the strength of the relationship between two variables.
The most commonly used correlation coefficient is the Pearson coefficient, which ranges from -1.0 to +1.0.
A positive correlation indicates two variables that tend to move in the same direction.
A negative correlation indicates two variables that tend to move in opposite directions.
A correlation coefficient of -0.8 or lower indicates a strong negative relationship, while a coefficient of -0.3 or lower indicates a very weak one.

In [91]:
# calculate correlation coefficient
corr_coef = calculate_correlation(DF6)
print(corr_coef)
corr_coef2 = calculate_correlation(DF7)
print(corr_coef2)
corr_coef3 = calculate_correlation(DF8)
print(corr_coef3)
corr_coef4 = calculate_correlation(DF9)
print(corr_coef4)
corr_coef5 = calculate_correlation(DF10)
print(corr_coef5)

(0.11266702379436973, 0.7013620094387367)
(0.16235214416896915, 0.5480146642608515)
(0.4248988906818991, 0.100888003345688)
(0.45228431713273787, 0.10442987481284179)
(0.1821655895100144, 0.5158283956843357)


### calculating co-relation for the entire set of years data 

In [87]:
DF_1 = pd.DataFrame()
DF_2 = pd.DataFrame()

dflst1 = [DF1,DF2,DF3,DF4,DF5]
dflst2 = [DF6,DF7,DF8,DF9,DF10]

for item in dflst1:
    DF_1 = pd.concat([DF_1,item])
    
for item2 in dflst2:
    DF_2 = pd.concat([DF_2,item2])
    

In [92]:
result1 = calculate_correlation(DF_1)

result2 = calculate_correlation(DF_2)

print("Result 1 is - ",result1)
print("Result 2 is - ",result2)

Result 1 is -  (-0.1402607892776991, 0.20042786331382934)
Result 2 is -  (0.26154757820785013, 0.023412482928931434)


## Concluding thoughts

The results for the co-relation coefficient calculated for specific years does provide us with a semi-consistent value range but due to lack of larger dataset, the p-calue is generally very high and does not allow us to draw a generalisable conclusion but we could still say few things from the results.

For the first visualisation done to compare the `Living conditions in prison` with `Governament expenditure`, we get a very very weak negative value for the coefficient which does imply that higher expenditure by the governament on maintining these public systems can be a factor in better living conditions in prison which means that they do not get over-populated. 

For the second visualisation done to compare the `Living conditions in prison` with `Suicides in prison` , we get a very week positive value for the coefficient which implies that higher suicides can be an indicator of overpopulated prisons.

To get better p_values for our coefficients we also calculated it after merging the datasets for all the years together and it did provide us better p_values as you can see above, especially for the second set of compared variables, we get a p_value if 0.02 which is an indicator that the co-efficient value is not a result of chance and can be considered somewhat generalisable.